util/src/main/java/tech/pegasys/artemis/util/backing/tree/TreeNode.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.util.backing.tree;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.function.Function;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.apache.tuweni.crypto.Hash;
+import org.jetbrains.annotations.NotNull;
+import tech.pegasys.artemis.util.backing.tree.TreeNodeImpl.RootImpl;
+
+/**
+ * Basic interface for Backing Tree node Backing Binary Tree concept for SSZ structures is described
+ * here: https://github.com/protolambda/eth-merkle-trees/blob/master/typing_partials.md#tree
+ *
+ * <p>Tree node is immutable by design. Any update on a tree creates new nodes which refer both new
+ * data nodes and old unmodified nodes
+ */
+public interface TreeNode {
+
+  static TreeNode createRoot(Bytes32 val) {
+    return new RootImpl(val);
+  }
+
+  /** Leaf node of a tree which contains 'bytes32' value */
+  interface Root extends TreeNode {
+
+    /** Returns node value */
+    Bytes32 getRoot();
+
+    @Override
+    default Bytes32 hashTreeRoot() {
+      return getRoot();
+    }
+
+    /**
+     * @param target generalized index. Should be equal to 1
+     * @return this node if 'target' == 1
+     * @throws IllegalArgumentException if 'target' != 1
+     */
+    @NotNull
+    @Override
+    default TreeNode get(long target) {
+      checkArgument(target == 1, "Invalid root index: %s", target);
+      return this;
+    }
+
+    @Override
+    default TreeNode update(long target, Function<TreeNode, TreeNode> nodeUpdater) {
+      checkArgument(target == 1, "Invalid root index: %s", target);
+      return nodeUpdater.apply(this);
+    }
+  }
+
+  /** Branch node of a tree */
+  interface Commit extends TreeNode {
```

Comment:
What do you think of:
```suggestion
  interface InternalTreeNode extends TreeNode {
```

Reply:
(https://github.com/PegaSysEng/teku/pull/1133#discussion_r387120744)

Reply:
@mbaxter then may be `BranchNode` would be better? That sounds to me as more widespread term

Reply:
sure - sounds good to me